Add ASync Methods

Tyler-OBrien opened the pull request:
This adds ASync Query Methods to QueryConnection.

The SendTimeout and ReceiveTimeout properties on the Socket only apply to Synchronous Send/Receive. There are ReceiveAsync and SendAsync method overloads on Socket in .NET Standard 2.1 that support CancellationToken, but that would mean dropping support for .Net Framework. For now, there is a slightly hacky workaround I used.

- Implemented ASync Methods
- Updated Example to include ASync Methods
- Updated Test Server to working one
- Added Tests for ASync methods and for timeout/failure
- Fixed errors with the Disconnect method, since this is all UDP based all we need to do is dispose the UDPClient to free the socket.

Tyler-OBrien committed:
Add ASync methods

Tyler-OBrien committed:
Add ASync example

Tyler-OBrien committed:
Add ASync Test, Update test server

Tyler-OBrien committed:
Update Async Test, Add Timeout test

Tyler-OBrien committed:
Fix timeouts not applying to ASync functions

Tyler-OBrien committed:
Fix disconnecting, it's not necessary to Disconnect with UDP (and disposing does the same action of freeing the socket)

Tyler-OBrien committed:
Use Disconnect in examples and tests to properly dispose socket

Florian2406 reviewed (changes requested):
Currently this project uses version and package release notes from the `Okolni.Source.Query.csproj` Can you increase the version and add the change notes? I would suggest it's version 1.2.0 - The version is needed twice in "Version" and "Fileversion"

I could add it myself if you want to. But tbh it's my first time actively using PRs in github so I don't know if that prevents us from merging :smile: - Or do you have a better suggestion for version management?

Tyler-OBrien committed:
Bump version to 1.2.0 - Added Async Methods

Tyler-OBrien reviewed (commented):
Sure, I bumped the version and added some change notes.

Btw, I mentioned this in the pull request, but there is a slight hack used to get Async Receive with timeouts working in .Net Standard 2.0. If you don't care about supporting .Net Framework, the library could move to .Net Standard 2.1 or even .Net 5, to get native support for Async Receive with cancellations. 

I found some other parsing issues with Port, SteamID, SourceTvPort/SourceTvName, Keywords, and GameID, caused by a minor issue with parsing the Extra Data Fields and the SteamID/GameID values being Ulongs (unsigned 64-bit ints) but being treated as uints. Not sure if you want me to push those changes with this pull request or make a separate one after.

Thanks for your library, it's made well and resolves some values like Rules where other .Net libraries have had issues.

Florian2406 commented:
Thanks, I mainly used it myself to get the rules, that's why many other issues didn't quite come through to me. If you want to contribute your fixes we could include them in this PR as well. Also I don't really think we need to support .NET Framework any further. So apparently you did a lot more research than I did with UDP - I've seen the link you referenced. Is .NET Standard 2.1 capable of native Async support? If it is we could go for .NET Standard 2.1. If we need .NET 5 we could also go for .NET 5. With the Async functionality, the bug fixes and the move for .NET Standard 2.1/.NET 5 we could bump the major version and get them all at once - Thanks a lot for your contributions btw!

Tyler-OBrien committed:
Switch to .Net Standard 2.1 for ASync Receive Support

Tyler-OBrien committed:
Use Native ASync Receive with Cancellation Token

Tyler-OBrien committed:
ByteReader: Add methods for Missing data types
- Added GetUShort, GetUint, GetLong, GetUlong
- GetLong now returns long instead of uint like before

Tyler-OBrien committed:
Update Response / Constant Data types to match https://developer.valvesoftware.com/wiki/Server_queries

Tyler-OBrien committed:
Parse correct data types

Tyler-OBrien committed:
Fix parsing of Extra Data Fields

Tyler-OBrien committed:
Update Package Release Notes

Tyler-OBrien committed:
Throw clear error if attempting to parse Obsolete GoldSource Response

Tyler-OBrien commented:
.Net Standard 2.1 has ASync Receive methods with CancellationToken overloads, and would let the library target .Net core 3.1 which isn't end of life yet, so for now I updated to that.

There are some other somewhat minor changes, mostly with the data types of various InfoResponse properties to bring them in line with the data types of the valve query protocol dev docs. 

Players, MaxPlayers, Bots, Witnesses fromInfoResponse, and Id in MultiPacketResponse were using larger data types than necessary. Port, SteamId, SourceTvPort, and GameID from InfoResponse were using data types too small. None of the too small properties were being parsed anyway besides GameID due to the parsing issue with Extra Data Fields, but worth noting as those are breaking changes.

One other minor thing: I saw that in the source there is support for MultiPacket Responses like for Goldsource or Source Servers, but there isn't support for parsing the Obsolete [GoldSource](https://developer.valvesoftware.com/wiki/Server_queries#:~:text=3.0.0.-,obsolete%20goldsource%20response,-GoldSource%20servers%20reply) Response. For now, I just added an error if a GoldSource response is received. 

I've been using this library to query all the servers of a few games, and haven't run into any issues with parsing other than the ones that are fixed with this.

Tyler-OBrien committed:
Query Methods may throw ArgumentException if they do not receive an expected response

Florian2406 commented:
Sorry for the long response time. I think the PR is fine. But I think it would be better to bump the major version now, what do you think?

Tyler-OBrien committed:
Update Major Version

Tyler-OBrien commented:
Sure, sounds good to me.
